tests/fakeClient.py

```diff
@@ -0,0 +1,119 @@
+import os
+import asyncio
+import uuid
+import socketio
+
+WORKSPACE_DIR = os.path.expanduser("~/ImJoyWorkspace")
+# read token from file if exists
+with open(os.path.join(WORKSPACE_DIR, ".token"), "r") as f:
+    token = f.read()
+
+client_id = str(uuid.uuid4())
+session_id = str(uuid.uuid4())
+url = "http://localhost:9527"
+
+NAME_SPACE = "/"
+
+test_plugin_config = {
+    "name": "test-plugin",
+    "type": "native-python",
+    "version": "0.1.12",
+    "api_version": "0.1.2",
+    "description": "This is a test plugin.",
+    "tags": ["CPU", "GPU", "macOS CPU"],
+    "ui": "",
+    "inputs": None,
+    "outputs": None,
+    "flags": [],
+    "icon": None,
+    "env": "conda create -n test-env python=3.6.7",
+    "requirements": "pip: numpy",
+    "dependencies": [],
+}
+
+
+class FakeClient:
+    def __init__(self, url, client_id, session_id, token):
+        self.engine_info = None
+        self.url = url
+        self.client_id = client_id
+        self.session_id = session_id
+        self.token = token
+
+    async def init(self):
+        sio = socketio.AsyncClient()
+        self.sio = sio
+
+        @sio.on("connect")
+        async def on_connect():
+            print("I'm connected!")
+            await self.sio.emit(
+                "register_client",
+                {
+                    "id": self.client_id,
+                    "token": self.token,
+                    "base_url": self.url,
+                    "session_id": self.session_id,
+                },
+                namespace=NAME_SPACE,
+                callback=self.on_registered,
+            )
+
+        @sio.on("message")
+        async def on_message(data):
+            print("I received a message!")
+
+        @sio.on("my message")
+        async def on_message(data):
+            print("I received a custom message!")
+
+        @sio.on("disconnect")
+        async def on_disconnect():
+            print("I'm disconnected!")
+
+        await sio.connect(self.url)
+        await asyncio.sleep(30)
```

Comment:
What will happen here?

Reply:
I haven't figure out how should we write test for async code, so basically we want to wait for the tests to complete. Here, we connect and it will trigger a series of event involving registering the client, init a plugin etc., during this time we don't want the client to shutdown.

Reply:
Tests should be fast, preferably without I/O.

I think we should think about making a mock framework with mock plugins that allow us to test the engine without I/O or at least run fast.

Reply:
I think thinking about using `asyncio.waitfor` with a timeout, this is what we do in the karma tests.

Reply:
I would try to use a fixture approach in pytest, where the fixture(s) set up the needed environment, then the test will run, then the fixture is dismantled automatically.

Reply:
There are plugins for pytest that support asyncio and async tests.

https://github.com/pytest-dev/pytest-asyncio

Reply:
Thanks. I now used pytest-asyncio